catalog/plugin/catalog-plugin-metadatabackup/src/main/java/org/codice/ddf/catalog/plugin/metadata/backup/MetadataBackupPlugin.java

```diff
@@ -0,0 +1,313 @@
+ * even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details. A copy of the GNU Lesser General Public License
+ * is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.catalog.plugin.metadata.backup;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessCreateItem;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessDeleteItem;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessRequest;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessResourceItem;
+import org.codice.ddf.catalog.async.data.api.internal.ProcessUpdateItem;
+import org.codice.ddf.catalog.async.plugin.api.internal.PostProcessPlugin;
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.FrameworkUtil;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceReference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ddf.catalog.data.BinaryContent;
+import ddf.catalog.data.Metacard;
+import ddf.catalog.plugin.PluginExecutionException;
+import ddf.catalog.transform.CatalogTransformerException;
+import ddf.catalog.transform.MetacardTransformer;
+
+/**
+ * The MetadataBackupPlugin asynchronously backs up Metacard Metadata to the file system.  It implements
+ * the PostProcessPlugin in order to maintain synchronization with the catalog (CRUD).
+ * <p>
+ * The root backup directory and subdirectory levels can be configured in the
+ * MetadataBackupPlugin section in the admin console.
+ */
+
+public class MetadataBackupPlugin implements PostProcessPlugin {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MetadataBackupPlugin.class);
+
+    private static final String OUTPUT_DIRECTORY_PROPERTY = "outputDirectory";
+
+    private static final String FOLDER_DEPTH_PROPERTY = "folderDepth";
+
+    private static final String KEEP_DELETED_METACARDS_PROPERTY = "keepDeletedMetacards";
+
+    private static final String METACARD_TRANSFORMER_ID_PROPERTY = "metacardTransformerId";
+
+    private static final Integer MAX_FOLDER_DEPTH = 4;
+
+    private static final Integer MIN_FOLDER_DEPTH = 0;
+
+    private Boolean keepDeletedMetacards = false;
+
+    private String metacardTransformerId;
+
+    private MetacardTransformer metacardTransformer;
+
+    private String outputDirectory;
+
+    private Integer folderDepth;
```

Comment:
maybe initialize to a non-null value? If it remains null and is passed to the method below it will NPE.